### PR TITLE
CAutomobile

### PR DESCRIPTION
**Pull request by aap**
helis still aren't right for some reason, needs debugging.
not 100% sure about planes and hover suspension either.

### CODE REVIEW

File: src/vehicles/Automobile.cpp [hunk 1]
```diff
@@ -654,16 +832,29 @@ CAutomobile::ProcessControl(void)
...
+			if(GetStatus() == STATUS_PLAYER && pHandling->Flags & HANDLING_IS_BUS){
+				if(m_nBusDoorTimerEnd == 0)
+					m_nBusDoorTimerEnd = 1000;
+				else if(m_nBusDoorTimerEnd > (uint32)CTimer::GetTimeStepInMilliseconds())
+					m_nBusDoorTimerEnd -= CTimer::GetTimeStepInMilliseconds();
+			}
+
+			if(m_aSuspensionSpringRatio[0] < 1.0f || m_aSuspensionSpringRatio[1] < 1.0f ||
+			   m_aSuspensionSpringRatio[2] < 1.0f || m_aSuspensionSpringRatio[3] < 1.0f)
+				ApplyTurnForce(-0.008*Min(m_fTurnMass, 2500.0f)*GetUp(), -1.0f*GetForward());
```
Comment (Nick007J) [inline src/vehicles/Automobile.cpp:849]: is it GRAVITY?

Comment (aap) [reply src/vehicles/Automobile.cpp:849]: probably

File: src/vehicles/Automobile.cpp [hunk 2]
```diff
@@ -3077,6 +3864,44 @@ CAutomobile::DoDriveByShootings(void)
...
+void
+CAutomobile::DoHoverSuspensionRatios(void)
+{
+	int i;
+
+	if(GetUp().z < 0.1f)
+		return;
+
+	CColModel *colmodel = GetColModel();
+	for(i = 0; i < 4; i++){
```
Comment (ShFil119) [inline src/vehicles/Automobile.cpp:3876]: `ARRAYSIZE`? for example `m_aSuspensionSpringRatio`

File: src/vehicles/Automobile.cpp [hunk 3]
```diff
@@ -3556,7 +4425,20 @@ CAutomobile::AddWheelDirtAndWater(CColPoint *colpoint, uint32 belowEffectSpeed)
 		for(i = 0; i < 4; i++){
 			dir.z = CGeneral::GetRandomNumberInRange(0.03f, 0.06f);
 			CParticle::AddParticle(PARTICLE_WHEEL_DIRT, colpoint->point, dir, nil,
-				CGeneral::GetRandomNumberInRange(0.02f, 0.06f), dirttrackCol);
+				CGeneral::GetRandomNumberInRange(0.02f, 0.06f), mudCol);
+		}
+		return 0;
+	case SURFACE_SAND:
+	case SURFACE_SAND_BEACH:
+		if(CTimer::GetFrameCounter() & 2 ||
```
Comment (ShFil119) [inline src/vehicles/Automobile.cpp:4433]: it would be `%3`?

File: src/vehicles/Automobile.cpp [hunk 4]
```diff
@@ -3556,7 +4425,20 @@ CAutomobile::AddWheelDirtAndWater(CColPoint *colpoint, uint32 belowEffectSpeed)
...
+		   CGeneral::GetRandomNumberInRange(CWeather::WetRoads, 1.01f) > 0.5f)
+			return 0;
+		dir.x = 0.5f*m_vecMoveSpeed.x;
+		dir.y = 0.5f*m_vecMoveSpeed.y;
+		for(i = 0; i < 1; i++){
```
Comment (ShFil119) [inline src/vehicles/Automobile.cpp:4438]: loop?

File: src/vehicles/Automobile.cpp [hunk 5]
```diff
@@ -4242,6 +5148,27 @@ CAutomobile::ProcessSwingingDoor(int32 component, eDoors door)
...
+		CObject *comp = SpawnFlyingComponent(CAR_BONNET, COMPGROUP_BONNET);
+#else
+		CObject *comp = SpawnFlyingComponent(CAR_BONNET, COMPGROUP_DOOR);
+#endif
+		// make both doors invisible on car
+		SetComponentVisibility(m_aCarNodes[CAR_BONNET], ATOMIC_FLAG_NONE);
+		Damage.SetDoorStatus(DOOR_BONNET, DOOR_STATUS_MISSING);
+
+		if(comp){
+			if(CGeneral::GetRandomNumber() & 1)
```
Comment (ShFil119) [inline src/vehicles/Automobile.cpp:5165]: `%2`

File: src/vehicles/Automobile.cpp [hunk 6]
```diff
@@ -4268,6 +5195,19 @@ CAutomobile::Fix(void)
 			mat.UpdateRW();
 		}
 	}
+
+	for(component = 0; component < 4; component++)
```
Comment (ShFil119) [inline src/vehicles/Automobile.cpp:5199]: ARRAYSIZE

File: src/vehicles/Automobile.cpp (hunk 4 above)
Comment (aap) [reply src/vehicles/Automobile.cpp:4438]: strange, but that's what the game is doing.

File: src/vehicles/Automobile.cpp (hunk 6 above)
Comment (aap) [reply src/vehicles/Automobile.cpp:5199]: Of what? i think the hardcoded 4 is fine. if you really want to change the number of wheels a car has you'll have to rewrite the whole class anyway

File: src/vehicles/Automobile.cpp (hunk 3 above)
Comment (aap) [reply src/vehicles/Automobile.cpp:4433]: i think they used bit operations and i find the code clearer too

**Pull request merged**